Download folders in .tar.gz format

boastful-squirrel opened the pull request:
This is my WIP for solving #2 (only .tar.gz at the moment).

The code isn't quite clean, the error handling needs some more work, and I have a bug on Mac OS X because user folders are wierd symlinks, and it causes the current code to fail. I'm investigating this issue. 

Otherwise, it works pretty well :)

boastful-squirrel commented:
So, I had to disable symlinks following strategy in the `TarBuilder` object because of these issues:

- https://github.com/alexcrichton/tar-rs/issues/147
- https://github.com/alexcrichton/tar-rs/issues/174

That's temporary, hopefully !

svenstaro commented on src/listing.rs:
```diff
@@ -218,17 +227,43 @@ pub fn directory_listing<S>(
+                );
+                Ok(HttpResponse::Ok()
+                    .status(http::StatusCode::INTERNAL_SERVER_ERROR)
+                    .body(""))
```
A helpful message here wouldn't hurt.

boastful-squirrel replied:
Yup, that's in my to-do list :)

boastful-squirrel replied:
I'm wondering, how precise do you want the messages to be ? Most descriptions from `std::io::Error` are not really helpful on the user-end, for instance `Is a directory` doesn't says much. Should I then go deeper than that ?

I know there's the `common_failures` crate that might help us in this case, at least for the `std::io::Error`...

svenstaro replied:
Well, they don't have to be too precise on the client side but it'd be helpful to have something more than an empty text. They should be pretty accurate on the server side of things, though.

svenstaro replied:
I don't think you ever gave this any text. It should have at least a somewhat rough message for what's wrong. Even if you just print the name of the bottom-most exception.

boastful-squirrel commented:
So, I started adding helpful messages for errors which might occur during the archiving of a folder. It's not done, I still miss some details on some errors, however I can print the full chain of errors which could be helpful for debugging if a user has a problem. I'm not quite sure about the messages I wrote, I'll try to improve them. Oh, and I'll add the missing details for some errors :)

boastful-squirrel commented:
Solving the tar-rs' issue made me realize that I don't have to actually create the temporary tar file, I can store the bytes in a buffer, just like I do with `gzip`. I'll implement that tonight :)

svenstaro commented:
I just had a super cursory view and it kind of appears like you're rebuilding pretty-env-logger. I think we should use standard Rust logging.

boastful-squirrel commented:
Right, I'll be on it ;)

boastful-squirrel commented:
I switched to `log`, it's clearer indeed. I set the level to `Error` if not in `verbose` mode because I think it's still important for a user to know what happened. 

Also, I changed the _"Info: miniserve has been invoked [...]"_ to an orange notice (kind of a warning, but not on the `log` stream), because I think it can be a security issue in some cases, so better really warn the user. I even thought about putting the whole message in orange, I don't know.

Also, my PR (https://github.com/alexcrichton/tar-rs/pull/187) is going good, I have to add some comments and we should be fine, then we can wait for the new release of `tar-rs`, and eventually we'll be able to set the `follow_symlink` option to `true` (`false` if `-P`) in `archive.rs` 👍

svenstaro commented:
Yes, it's pretty big already! Let's keep it .tar.gz.

boastful-squirrel commented:
Great! I'll keep you posted when it comes to the fix of tar-rs.

svenstaro commented:
It's amazing that you're fixing this issue and the two over at tar-rs in
one go! Keep it up.

On Thu, Mar 14, 2019, 18:56 Boastful Squirrel <notifications@github.com>
wrote:

> Great! I'll keep you posted when it comes to the fix of tar-rs.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/svenstaro/miniserve/pull/48#issuecomment-472821672>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGgApLnYZmOvKUwYyKtBkp24-ZFkhKks5vWjjkgaJpZM4bmKAA>
> .
>

boastful-squirrel commented:
:)

So my PR has been merged, hopefully a new version will be released

boastful-squirrel commented:
Here we go ;)

svenstaro commented on src/renderer.rs:
```diff
@@ -259,6 +275,25 @@ fn css() -> Markup {
         text-decoration: none;
     }
+    .download {
+        display: flex;
```
I kind of don't like how this pushes the rest of the content away quite a bit. How about instead position the download button right of the header where it tells you the current directory?

boastful-squirrel replied:
You're right. I did some experiments and it's kind of hard to make it look right with very long path though. How about just putting on a new line, but align to the left ? Here's a screenshot, with 1 and then several download buttons

<img width="1335" alt="Screenshot 2019-03-15 at 17 22 21" src="https://user-images.githubusercontent.com/47121073/54446685-ced7a580-4747-11e9-9cf1-eb48790da534.png">

<img width="1340" alt="Screenshot 2019-03-15 at 17 23 02" src="https://user-images.githubusercontent.com/47121073/54446659-c1222000-4747-11e9-9ffc-2be94f838840.png">

svenstaro commented on src/errors.rs:
```diff
@@ -0,0 +1,90 @@
+
+/// Kinds of errors which might happen during the generation of an archive
+#[derive(Debug, Fail)]
+pub enum CompressionErrorKind {
```
While this seems to have a lot of boilerplate, I'm afraid it's the correct choice for this application.

boastful-squirrel replied:
So should I fix something on that aspect ?

svenstaro replied:
Nah, let's leave this part as is.

svenstaro commented on src/archive.rs:
```diff
@@ -0,0 +1,138 @@
+    }
+}
+
+pub fn create_archive_file(
```
Document this.

svenstaro commented on src/archive.rs:
```diff
@@ -0,0 +1,138 @@
+    }
+}
+
+/// Compresses a given folder in .tar.gz format
```
Same, please document inputs/outputs.

boastful-squirrel commented:
I edited the documentation so it's a bit clearer, also I implemented the design idea I posted https://github.com/svenstaro/miniserve/pull/48#discussion_r266056153. I'm available for fixes if needed.

As for your last comment, I'll take a look at this, I have no idea wether it's possible, but that sounds like a cool challenge. I didn't really see the problem, since the compression job does not block the user from browsing the server.

svenstaro commented:
The web server should start sending bytes as soon as we give it bytes but I don't we give it bytes early enough. I'm pretty sure it should work. I'm unsure that what you're doing right now is even supposed to work: Setting the `Content-Length` and `chunked` at the same time. See https://gist.github.com/CMCDragonkai/6bfade6431e9ffb7fe88.

Basically, while we're archiving, we have no idea how long the entire file is going to be but this behavior is already supported by all browsers, thankfully.

boastful-squirrel commented:
Ok, I get what you mean. So, the first step will probably to make all the methods in `archive.rs` non blocking, they should stream the results of bytes, instead of waiting for the whole input to be consumed. Well, that's gotta be fun ! :)

svenstaro commented:
I never said it was gonna be easy but in the end I think the payoff is going to be worth it since it will be a lot more useful for sharing large directories! :)

svenstaro commented:
Keep at it, you've done a marvelous job so far!

boastful-squirrel commented:
I fear the tar-rs crate won't let us stream its result... All methods seem blocking, I'll dig deeper in the doc

svenstaro commented:
Ok that really is a bummer. Tell you what, I'll merge this as is without async but I really kinda want async in tbe long run because it does enable sending really large directories which currently is impossible.

boastful-squirrel commented:
I agree with you. Maybe we should add a little mention in the README, so people don't start trying to share huge folders

boastful-squirrel commented:
Maybe I could try to reimplement the method that browses a directory, and stream the result for each file...

svenstaro commented:
That seems like band-aid for what should really be an implementation detail of tar-rs. As I said, we'll need to fix this properly in tar-rs if anything. It shouldn't be _that_ hard.

svenstaro commented:
Please take a look at the remaining open comments.

boastful-squirrel commented:
I think the only comment I didn't fix is https://github.com/svenstaro/miniserve/pull/48#discussion_r266048905, however I don't know if it requires a fix, or if it was just a comment

svenstaro commented:
Can you just https://github.com/svenstaro/miniserve/pull/48#discussion_r266044603 a bit of text and that's that?

boastful-squirrel commented:
No idea what I broke :/

boastful-squirrel commented:
I added a little warning in the readme about the compression not being async-ready. I hope that's fine with you.

svenstaro commented:
Ok, back from vacation and merging! Stellar work. Hopefully we can make this async at some point.
